plugins/user-authenticators/saml2/src/org/apache/cloudstack/saml/SAMLTokenDaoImpl.java

```diff
@@ -27,17 +29,22 @@
 @DB
 @Component
 public class SAMLTokenDaoImpl extends GenericDaoBase<SAMLTokenVO, Long> implements SAMLTokenDao {
+    protected final SearchBuilder<SAMLTokenVO> SessionIndexSearchNotSpBaseUrl;
 
     public SAMLTokenDaoImpl() {
         super();
+        SessionIndexSearchNotSpBaseUrl = createSearchBuilder();
+        SessionIndexSearchNotSpBaseUrl.and("session_index", SessionIndexSearchNotSpBaseUrl.entity().getSessionIndex(), SearchCriteria.Op.EQ);
+        SessionIndexSearchNotSpBaseUrl.and("sp_base_url", SessionIndexSearchNotSpBaseUrl.entity().getSpBaseUrl(), SearchCriteria.Op.NEQ);
+        SessionIndexSearchNotSpBaseUrl.done();
     }
 
     @Override
     public void expireTokens() {
         TransactionLegacy txn = TransactionLegacy.currentTxn();
         try {
             txn.start();
-            String sql = "DELETE FROM `saml_token` WHERE `created` < (NOW() - INTERVAL 1 HOUR)";
+            String sql = "DELETE FROM `saml_token` WHERE `created` < (NOW() - INTERVAL 1 HOUR) AND session_index IS NULL OR `created` < (NOW() - INTERVAL 24 HOUR) AND session_index IS NOT NULL";
```

Comment:
Are you certain your AND/OR logic is doing exactly what you intend? The inclusion of that OR in the middle, with an AND following it makes me unsure of the order. I'd have to look it up. Maybe add some parens to clarify?